test/SILOptimizer/utf8_decoding_fastpath.swift

```diff
@@ -0,0 +1,79 @@
+// RUN: %target-swift-frontend -parse-as-library -O -target-cpu core2 -emit-ir  %s | %FileCheck %s
+// REQUIRES: optimized_stdlib,CPU=x86_64
+
+// This is an end-to-end test to ensure that the optimizer generates
+// good code for various ways to create a String from a Sequence of UTF-8
+// bytes for which a fast path exists.
+
+// Please note: this test targets "core2" to ensure consistent output
+// on all x86 host processors.
+
+// UnsafeBufferPointer<UInt8>: we check that we call the merged function
+// ==========================
+// CHECK-LABEL: define swiftcc {{.*}}s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGF{{.*}}
+// CHECK-NEXT: {{.*}} = tail call swiftcc {{.*}}s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGFTm{{.*}}
+// CHECK: ret
+public func decodeUBPAsUTF8(_ ptr: UnsafeBufferPointer<UInt8>) -> String {
+  return String(decoding: ptr, as: Unicode.UTF8.self)
+}
+
+// UnsafeMutableBufferPointer<UInt8>: we check that we call the merged function
+// =================================
+// CHECK-LABEL: define swiftcc {{.*}}s22utf8_decoding_fastpath16decodeUMBPAsUTF8ySSSrys5UInt8VGF{{.*}}
+// CHECK-NEXT: {{.*}} = tail call swiftcc {{.*}}s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGFTm{{.*}}
+// CHECK: ret
+public func decodeUMBPAsUTF8(_ ptr: UnsafeMutableBufferPointer<UInt8>) -> String {
+  return String(decoding: ptr, as: Unicode.UTF8.self)
+}
+
+// this defined the merged function that handles both UBP and UMBP
+// CHECK-LABEL: define internal swiftcc {{.*}}s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGFTm{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: {{.*}} = call swiftcc {{.*}} @"$sSS18_fromUTF8Repairing{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: ret
+
```

Comment:
Interesting. @eeckstein, is there a more robust way to do this?

Reply:
yeah, I wasn't happy with this either but thought better than nothing... @eeckstein ?